onnxruntime/core/providers/cuda/tensor/gather.cc

```diff
@@ -87,6 +105,102 @@ Status Gather::ComputeInternal(OpKernelContext* context) const {
+        IAllocatorUniquePtr<SegmentIndex_t> per_segment_partial_segment_counts_out;
+        IAllocatorUniquePtr<SegmentIndex_t> per_segment_partial_segment_offsets_out;
+
+        if (p.indices_tensor->IsDataType<int32_t>())
+        {
+            IAllocatorUniquePtr<int32_t> dX_indices_sorted_out, dY_indices_sorted_out;
+            GatherGradPrepare<int32_t>(
+                Stream(),
+                CudaScratchBufferAllocator{*this},
+                reinterpret_cast<const int32_t*>(indices_data),
+                num_gathered_indices,
+                gather_dimension_size,
+                num_gathered_per_index,
+                *p_num_segments, segment_offsets_out,
+                last_segment_partial_segment_count_out,
+                last_segment_partial_segment_offset_out,
+                per_segment_partial_segment_counts_out,
+                per_segment_partial_segment_offsets_out,
+                dX_indices_sorted_out,
+                dY_indices_sorted_out);
+
+            auto* dX_indices_sorted = context->Output(7, p.indices_tensor->Shape());
+            int32_t* p_dX_indices_sorted = dX_indices_sorted->MutableData<int32_t>();
+            CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(p_dX_indices_sorted, dX_indices_sorted_out.get(),
+                                                    dX_indices_sorted->SizeInBytes(), cudaMemcpyDeviceToDevice, Stream()));
+
+            auto* dY_indices_sorted = context->Output(8, p.indices_tensor->Shape());
+            int32_t* p_dY_indices_sorted = dY_indices_sorted->MutableData<int32_t>();
+            CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(p_dY_indices_sorted, dY_indices_sorted_out.get(),
+                                                    dY_indices_sorted->SizeInBytes(), cudaMemcpyDeviceToDevice, Stream()));
+        } else
+        {
+            IAllocatorUniquePtr<int64_t> dX_indices_sorted_out, dY_indices_sorted_out;
+            GatherGradPrepare<int64_t>(
+                Stream(),
+                CudaScratchBufferAllocator{*this},
+                reinterpret_cast<const int64_t*>(indices_data),
+                num_gathered_indices,
+                gather_dimension_size,
+                num_gathered_per_index,
+                *p_num_segments, segment_offsets_out,
+                last_segment_partial_segment_count_out,
+                last_segment_partial_segment_offset_out,
+                per_segment_partial_segment_counts_out,
+                per_segment_partial_segment_offsets_out,
+                dX_indices_sorted_out,
+                dY_indices_sorted_out);
+
+            auto* dX_indices_sorted = context->Output(7, p.indices_tensor->Shape());
+            int64_t* p_dX_indices_sorted = dX_indices_sorted->MutableData<int64_t>();
+            CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(p_dX_indices_sorted, dX_indices_sorted_out.get(),
+                                                    dX_indices_sorted->SizeInBytes(), cudaMemcpyDeviceToDevice, Stream()));
```

Comment:
I think all the D2D memcpy in this function can be avoided.... 

For example, dX_indices_sorted can be created first, and then passed into GatherGradPrepare function.

Reply:
In order for me to remove all the DtoDs, I had to split `GatherGradPrepare` into two parts since a few outputs depend on the number of segments in the sorted indices. Let me know if you think it is too verbose and if we should probably simplify it further.